src/storages/auth.storage.js

```diff
@@ -1,85 +1,159 @@
+
+
+    if (!firstName || !lastName || !email || !password)
+        return res.status(400).json({ auth: false, message: "some fields are mandatory" });
+
+    const validRegex = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+    if (!email.match(validRegex)) {
+        return res.status(400).json({ auth: false, message: "Invalid Email" });
+    }
+    User.findOne({ email: newuser.email }, function (err, user) {
+        if (user) return res.status(400).json({ auth: false, message: "email exists" });
+        
+        newuser.save((err, doc) => {
+            if (err) {
+                console.log(err);
+                return res.status(400).json({ success: false });
+            }
+            res.status(200).json({
+                succes: true,
+                user: doc,
+                message: "User was registered successfully! Please check your email",
+            });
+
+
+            nodemailer.sendConfirmationEmail(
+                newuser.firstName,
+                newuser.email,
+                newuser.emailToken
+            );
+            console.log("before email not verified ",newuser.isVerified)
+        });
+    });
+
+    //res.redirect('/user/login')
+
 };
 
+const verifyUser = (req, res, next) => {
+    User.findOne({
+        confirmationCode: req.params.emailToken,
+    })
+        .then((user) => {
+            if (!user) {
+                return res.status(404).send({ message: "User Not found." });
+            }
+            user.isVerified=true
+            
+            //user.isVerified=true;
+            console.log(user.isVerified + " User is verified after the click on email")
+            user.save((err) => {
+                if (err) {
+                    res.status(500).send({ message: err });
+                    return;
+                }
+                
+                
+            });
+            /*NOTA BENE : At this moment, if the user clicks on the email’s confirmation link, they will find
+             a blank page and still be unable to log in. Therefore, we need to make some changes on the front 
+             end to complete the registration procedure. See https://betterprogramming.pub/how-to-create-a-signup-confirmation-email-with-node-js-c2fea602872a*/
+            //res.redirect('./api/login')
+        })
+        .catch((e) => console.log("error", e));
+}
 
 // login user
-    const login=(req,res)=>{
-    let token=req.cookies.auth;
-    User.findByToken(token,(err,user)=>{
-        if(err) return  res(err);
-        if(user) return res.status(400).json({
-            error :true,
-            message:"You are already logged in"
+const login = (req, res) => {
+    let token = req.cookies.auth;
+    User.findByToken(token, (err, user) => {
+        if (err) return res(err);
```

Comment:
Quale codice ritorna all'utente in questa risposta?

Reply:
praticamente cerca il token di autenticazione, se è già presente vuol dire che l'utente è già loggato, altrimenti genera il token e viene fatto il login (facendo prima tutti i controlli, tipo se ti logghi con una email non presente, la password non corrisponde etc) #